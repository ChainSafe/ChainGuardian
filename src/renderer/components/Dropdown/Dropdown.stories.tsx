import * as React from "react";
import {useState} from "react";
import {storiesOf} from "@storybook/react";
import {withKnobs, array} from "@storybook/addon-knobs";
import {Dropdown} from "./Dropdown";

storiesOf("Dropdown", module)
    .add("Dropdown", () => {
        const options = array("options", ["All networks", "Mainnet", "Testnet", "Networkname #1"]);
        const [current, setCurrent] = useState(0);
        return (
            <div style={{height: "500px"}}>
                <Dropdown current={current} onChange={(selected): void => setCurrent(selected)} options={options} />
            </div>
        );
    })
    .addDecorator(withKnobs);
