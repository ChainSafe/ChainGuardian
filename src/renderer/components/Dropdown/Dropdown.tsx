import * as React from "react";
import {ReactNode, useState} from "react";
import cgLogo from "../../assets/ico/app_icon.png";

export interface IDropdownProps {
    options: Array<string> | {[id: number]: string};
    current: number;
    label?: string;
    onChange?: (selected: number) => void;
    verifiedIndex?: number;
    className?: string;
}

export const Dropdown: React.FunctionComponent<IDropdownProps> = ({
    options,
    current,
    label,
    onChange,
    verifiedIndex,
    className = "",
}) => {
    const [visible, setVisible] = useState("none");
    // eslint-disable-next-line no-param-reassign
    options = Array.isArray(options) ? {...options} : options;

    function showHide(): void {
        visible === "none" ? setVisible("block") : setVisible("none");
    }

    function hide(): void {
        visible === "block" ? setVisible("none") : null;
    }

    function onOptionClick(key: number): void {
        if (onChange) {
            onChange(key);
            showHide();
        }
    }

    function renderOption(key: number): ReactNode {
        return (
            <div
                key={options[key]}
                onClick={(): void => onOptionClick(key)}
                className={`dropdown-item
                ${visible}
                ${key === current ? "selected" : ""}`}>
                {options[key]}
                {verifiedIndex === key && <img alt='ChainGuardian Logo' src={cgLogo} />}
            </div>
        );
    }

    return (
        <div className={className}>
            {label && <h3>{label}</h3>}
            <div onClick={(): void => hide()} className='dropdown-screen'>
                <div className='dropdown-container'>
                    <div onClick={(): void => onChange && showHide()} className='dropdown-selected'>
                        <div>{options[current]}</div>
                    </div>
                    <div className='dropdown-items-container'>
                        <div className='dropdown-items'>
                            {Object.keys(options)
                                .map((v) => parseInt(v))
                                .map((key) => renderOption(key))}
                        </div>
                    </div>
                </div>
            </div>
        </div>
    );
};
