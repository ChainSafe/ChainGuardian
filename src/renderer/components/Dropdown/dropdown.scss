.dropdown-screen{
    // position: absolute;
    //width: 2000px;
    //height: 2000px;
    // z-index: -100;
}
.dropdown-container{
display: flex;
flex-direction: column;
font-family: "Palanquin", sans-serif;
font-weight: 400;
font-size: 18px;
line-height: 28px;
color: $neutral-100;
z-index: 100;
}

.dropdown-selected{
    background-color: $neutral-600;
    margin-bottom: 8px;
    padding: 13px 56px 13px 24px;
    width: 252px;
    border-radius: 8px;
    z-index: 200;
    position: relative;

    &::after {
        content: url(../assets/img/dropdown/Expand.svg);
        position: absolute;
        top: 30%;
        right: 20px;
    }
}
.dropdown-selected:hover{
    background-color: $neutral-600;
    cursor: pointer;
}

.dropdown-items-container{
    position: relative;
    width: 0;
    height: 0;
    z-index: 300;
}

.dropdown-items{
    position: absolute;
    z-index: 100;
    display: flex;
    flex-direction: column;
    background-color: fade-out($neutral-700, 0.01);
    width: 332px;
    border-radius: 8px;
    transform: translateY(-15px);
    // position: absolute;
}
.dropdown-item{
    // display: none;
    border-bottom: 1px solid $neutral-600;
    margin-bottom: 8px;
    padding: 13px 24px;
    margin: 0px;
    &.none{
        display: none;
    }
    &.block{
        display: block;
    }
    &.selected{
        background-color: $neutral-600;
    }

    img {
        height: 18px;
        margin-left: 10px;
        margin-bottom: -3px;
    }
}
.dropdown-item:first-child{
    border-radius:8px 8px 0 0;
}
.dropdown-item:last-child{
    border: none;
    border-radius:0 0 8px 8px;
}
.dropdown-item:hover{
    background-color: $neutral-600;
    cursor: pointer;
}