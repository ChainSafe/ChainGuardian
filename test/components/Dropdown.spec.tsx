import * as React from "react";
import * as renderer from "react-test-renderer";
import {Dropdown} from "../../src/renderer/components/Dropdown/Dropdown";

describe("Dropdown", () => {
    it("renders correctly with array of options", () => {
        const tree = renderer
            .create(
                <Dropdown
                    onChange={(): void => {}}
                    current={0}
                    options={["All networks", "Mainnet", "Testnet", "Networkname #1"]}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders correctly with options object", () => {
        const tree = renderer
            .create(
                <Dropdown
                    onChange={(selected): void => console.log(selected)}
                    current={0}
                    options={{0: "All networks", 112: "Mainnet", 23: "Testnet", 41: "Networkname #1"}}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
