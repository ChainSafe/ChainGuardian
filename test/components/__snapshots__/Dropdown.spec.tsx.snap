// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly with array of options 1`] = `
<div
  className=""
>
  <div
    className="dropdown-screen"
    onClick={[Function]}
  >
    <div
      className="dropdown-container"
    >
      <div
        className="dropdown-selected"
        onClick={[Function]}
      >
        <div>
          All networks
        </div>
      </div>
      <div
        className="dropdown-items-container"
      >
        <div
          className="dropdown-items"
        >
          <div
            className="dropdown-item
                none
                selected"
            onClick={[Function]}
          >
            All networks
          </div>
          <div
            className="dropdown-item
                none
                "
            onClick={[Function]}
          >
            Mainnet
          </div>
          <div
            className="dropdown-item
                none
                "
            onClick={[Function]}
          >
            Testnet
          </div>
          <div
            className="dropdown-item
                none
                "
            onClick={[Function]}
          >
            Networkname #1
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown renders correctly with options object 1`] = `
<div
  className=""
>
  <div
    className="dropdown-screen"
    onClick={[Function]}
  >
    <div
      className="dropdown-container"
    >
      <div
        className="dropdown-selected"
        onClick={[Function]}
      >
        <div>
          All networks
        </div>
      </div>
      <div
        className="dropdown-items-container"
      >
        <div
          className="dropdown-items"
        >
          <div
            className="dropdown-item
                none
                selected"
            onClick={[Function]}
          >
            All networks
          </div>
          <div
            className="dropdown-item
                none
                "
            onClick={[Function]}
          >
            Testnet
          </div>
          <div
            className="dropdown-item
                none
                "
            onClick={[Function]}
          >
            Networkname #1
          </div>
          <div
            className="dropdown-item
                none
                "
            onClick={[Function]}
          >
            Mainnet
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
